src/main/java/org/passay/RuleResultMetadata.java

```diff
@@ -0,0 +1,89 @@
+/* See LICENSE for licensing and NOTICE for copyright. */
+package org.passay;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * Describes metadata relevant to the result of rule validation.
+ *
+ * @author  Middleware Services
+ */
+public class RuleResultMetadata
+{
+
+  /** Additional metadata that provide information about the rule. */
+  protected final Map<String, Object> metadata;
```

Comment:
I can appreciate the flexibility of a map of objects, but it deprives clients of knowing about what should be well-defined metadata categories at compile time. I'm in favor of getters for the categories, but maybe we can split the difference with an enum for the categories. Then you could have put methods for both well-known categories by enum as well as ad hoc categories by string. In that view the underlying map could still be keyed on a string.

Reply:
Concrete metadata categories places a high burden on some of the rules to implement logic to determine which category metadata falls into. See `CharacterRule`. I didn't want metadata code bleeding into the `CharacterData` classes, but that would be likely result of this sort of change.

Reply:
Actually, if we're willing to drive this via naming convention...I could make the category names line up with the `EnglishCharacterData` enums in such a way to make this work.
I'll investigate.

Reply:
That approach sounds promising.

Reply:
Pushed a change for this.